go/interpreter/lfvm/instruction_statistics.go

```diff
@@ -0,0 +1,180 @@
+// number of times each pair, triple, and quad of instructions are executed.
+type statistics struct {
+	count       uint64
+	singleCount map[uint64]uint64
+	pairCount   map[uint64]uint64
+	tripleCount map[uint64]uint64
+	quadCount   map[uint64]uint64
+}
+
+func newStatistics() *statistics {
+	return &statistics{
+		singleCount: map[uint64]uint64{},
+		pairCount:   map[uint64]uint64{},
+		tripleCount: map[uint64]uint64{},
+		quadCount:   map[uint64]uint64{},
+	}
+}
+
+// insert adds the instruction counts of the given statistics to this instance.
+func (s *statistics) insert(src *statistics) {
+	s.count += src.count
+	for k, v := range src.singleCount {
+		s.singleCount[k] += v
+	}
+	for k, v := range src.pairCount {
+		s.pairCount[k] += v
+	}
+	for k, v := range src.tripleCount {
+		s.tripleCount[k] += v
+	}
+	for k, v := range src.quadCount {
+		s.quadCount[k] += v
+	}
+}
+
+// print returns a human-readable summary of the collected statistics.
+func (s *statistics) print() string {
+
+	type entry struct {
+		value uint64
+		count uint64
+	}
+
+	getTopN := func(data map[uint64]uint64, n int) []entry {
+		list := make([]entry, 0, len(data))
+		for k, c := range data {
+			list = append(list, entry{k, c})
+		}
+		sort.Slice(list, func(i, j int) bool {
+			return list[i].count > list[j].count
+		})
+		if len(list) < n {
+			return list
+		}
+		return list[0:n]
+	}
+
+	builder := strings.Builder{}
+	write := func(format string, args ...interface{}) {
+		builder.WriteString(fmt.Sprintf(format, args...))
+	}
+
+	write("\n----- Statistics ------\n")
+	write("\nSteps: %d\n", s.count)
+	write("\nSingles:\n")
+	for _, e := range getTopN(s.singleCount, 5) {
+		write("\t%-30v: %d (%.2f%%)\n", OpCode(e.value), e.count, float32(e.count*100)/float32(s.count))
+	}
+	write("\nPairs:\n")
+	for _, e := range getTopN(s.pairCount, 5) {
+		write("\t%-30v%-30v: %d (%.2f%%)\n", OpCode(e.value>>16), OpCode(e.value), e.count, float32(e.count*100)/float32(s.count))
+	}
+	write("\nTriples:\n")
+	for _, e := range getTopN(s.tripleCount, 5) {
+		write("\t%-30v%-30v%-30v: %d (%.2f%%)\n", OpCode(e.value>>32), OpCode(e.value>>16), OpCode(e.value), e.count, float32(e.count*100)/float32(s.count))
+	}
+
+	write("\nQuads:\n")
+	for _, e := range getTopN(s.quadCount, 5) {
+		write("\t%-30v%-30v%-30v%-30v: %d (%.2f%%)\n", OpCode(e.value>>48), OpCode(e.value>>32), OpCode(e.value>>16), OpCode(e.value), e.count, float32(e.count*100)/float32(s.count))
+	}
+	write("\n")
+
+	return builder.String()
+}
+
+// statsCollector is a helper struct that keeps track of the resent history of
+// instructions executed by the VM to collect instruction sequence statistics.
+type statsCollector struct {
+	stats *statistics
+
+	last       uint64
+	secondLast uint64
+	thirdLast  uint64
+}
+
+func (s *statsCollector) nextOp(op OpCode) {
+	if op > 255 {
+		panic("Instruction sequence statistics does not support opcodes > 255")
+	}
```

Comment:
The limit of 255 is not related to the number of OpCodes but is a limit of the implementation of the statistic type itself. It encodes sequences of up to 4 operations in a single 32-bit unsigned integer, where each operation can only occupy up to 8 bit -- thus the limit. It would be wrong to use the NUM_OPCODES constant here.